crates/database/src/fee_policies.rs

```diff
@@ -0,0 +1,170 @@
+use {
+    crate::{auction::AuctionId, OrderUid, PgTransaction},
+    bigdecimal::BigDecimal,
+    sqlx::PgConnection,
+    std::ops::DerefMut,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct FeePolicy {
+    pub auction_id: AuctionId,
+    pub order_uid: OrderUid,
+    pub kind: FeePolicyKind,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum FeePolicyKind {
+    PriceImprovement {
+        price_improvement_factor: f64,
+        max_volume_factor: Option<f64>,
+        max_absolute_fee: Option<BigDecimal>,
+    },
+    Volume {
+        factor: f64,
+    },
+}
+
+#[derive(Debug, Clone, PartialEq, sqlx::FromRow)]
+struct FeePolicyRow {
```

Comment:
What's the advantage of properly structuring the kind into columns? I think this will just make it very challenging to migrate. Would it make sense to instead store json-ified versions of the fee policy so we can add new policies without a db migration?

Reply:
@MartinquaXD what do you think here? I don't fully get the argument about db space (I'm not sure storing 3 values is more/less space efficient than a potentially compressed json blob, but in any case the difference is likely not meaningful for us).

I think we should think about maintainability (ie as we define more fee types is it better to have them as potentially formatted jsons or do we want to migrate the DB on each change (I don't really foresee advantages of the querying capacities, ie. I don't think we need to e.g. query all orders with absolute fee > some value)

Reply:
> So the advantage would be that we use less db space, since we have an entry for each auction for each order.

We only really need to store the data for orders that were settled in an auction, right?
In that case we should rather the format that fits our needs best and add a clean up task that deletes data that is no longer needed.

> I think we should think about maintainability (ie as we define more fee types is it better to have them as potentially formatted jsons or do we want to migrate the DB on each change

This would be the reason for me why preferring regular columns over JSON blobs make sense. One can easily imagine a scenario where you would want to look up some historic data which no longer works because the JSON format changed in an incompatible way in the mean time. When we have real columns we are forced to more conscious about keeping the system coherent. That way we can still chose to phase out old incompatible formats without being surprised all of a sudden.

Reply:
Right, but FeePolicy to me is something very "atomic" on the order which we are not likely to sub-query for specific ranges (buy I might by wrong).
In this case, I'd strongly argue to not overload the semantics of each column though (if volume fees have a volume % and quote deviation orders have a volume fee cap in %, let's define separate columns for both). No developer in their right mind is going to understand in 6 months time under what circumstances a value in one column affects the semantics of a value in another column.

Reply:
Switched to storing all columns and not overloading a single column for multiple things.

Reply:
Isn't "factor" for the volume variant still missing (or implicitly overloading "volume factor")?

Just as a thought experiment, how would we go about adding a new FeePolicyKind variant later on?

Reply:
Updated.

> Just as a thought experiment, how would we go about adding a new FeePolicyKind variant later on?

We would add a new enum variant to the kind and added new columns specific to this kind.